File: arrays-hashing/0001-two-sum/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>


template <typename T>
std::ostream& operator<<(std::ostream& os, const std::vector<T>& v) {
    os << "[";
    for (size_t i = 0; i < v.size(); i++) {
        os << v[i];
        if (i + 1 < v.size()) os << ", ";
    }
    os << "]";
    return os;
}

class Solution {
public:
    std::vector<int> twoSum(std::vector<int>& nums, int target) {
        std::unordered_map<int, int> hm;

        for (int i=0; i<nums.size(); i++) {
            int num = nums[i];
            if (hm.count(num)) 
                return std::vector<int> {hm[num], i};
            else 
                hm[target-num] = i;
        }
        return std::vector<int> {-1}; 
    }
};

struct TestCase {
    std::vector<int> nums;
    int target;
    std::vector<int> expected;
};

int main() {
    std::vector<TestCase> tests = {
        {{2, 7, 11, 15}, 9, {0, 1}},
        {{3, 2, 4}, 6, {1, 2}},
        {{3, 3}, 6, {0, 1}}
    };

    Solution sol;
    int correct = 0;

    for (size_t i = 0; i < tests.size(); i++) {
        TestCase& t = tests[i];
        auto actual = sol.twoSum(t.nums, t.target);
        bool passed = (actual == t.expected);
        correct += passed;

        std::cout << "Test " << i << ": " << (passed ? "PASS" : "FAIL") << "\n";
        std::cout << "\tExpected: " << t.expected << "\n";
        std::cout << "\tActual:   " << actual << "\n";
    }

    std::cout << "\n" << correct << " / " << tests.size() << " correct.\n";
    return 0;
}

File: arrays-hashing/0001-two-sum/solution.java
import java.util.*;

class Solution {
    public List<Integer> solve(List<Integer> nums, int target) {
      // Turns out leetcode uses int[] and not List<Integer>
      // Swaps documented in code
      Map<Integer, Integer> hm = new HashMap<>();
      for (int i=0; i<nums.size(); i++) { // .size() becomes .length
        int num = nums.get(i); // this becomes a straight index, nums[i]
        if (hm.containsKey(num)) {
          return List.of(hm.get(num), i); // new int[] {hm.get(num), i}
        } else {
          hm.put(target-num, i);
        }
      }
      return List.of(-1,-1); // new int[] {-1}
    }

    record TestCase(List<Integer> nums, int target, List<Integer> expected) {}

    public static void main(String[] args) {
        Solution s = new Solution();

        List<TestCase> tests = List.of(
            new TestCase(List.of(2,7,11,15), 9, List.of(0,1)),
            new TestCase(List.of(3,2,4), 6, List.of(1,2)),
            new TestCase(List.of(3,3), 6, List.of(0,1))
        );

        int correct = 0;
        for (int i = 0; i < tests.size(); i++) {
            TestCase t = tests.get(i);
            List<Integer> actual = s.solve(t.nums(), t.target());
            boolean passed = actual.equals(t.expected());
            correct += passed ? 1 : 0;

            System.out.printf("Test %d: %s%n", i, passed ? "PASS" : "FAIL");
            System.out.printf("\tExpected: %s%n", t.expected());
            System.out.printf("\tActual:   %s%n", actual);
        }

        System.out.printf("%n%d / %d correct.%n", correct, tests.size());
    }
}
